src/Microsoft.AspNet.OData.Shared/Builder/IODataInstanceAnnotationContainer.cs

```diff
@@ -0,0 +1,56 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.AspNet.OData.Builder
+{
+    /// <summary>
+    /// Interface to used as a Container for holding Instance Annotations, An default implementation is provided
+    /// Custoer can implement the interface and can have their own implementation.
+    /// </summary>
+    public interface IODataInstanceAnnotationContainer
+    {
+        /// <summary>
+        /// Method to Add an Instance Annotation to the CLR type
+        /// </summary>
+        /// <param name="annotationName">Name of Annotation</param>
+        /// <param name="value">Value of Annotation</param>
+        void AddResourceAnnotation(string annotationName, object value);
+
+        /// <summary>
+        /// Method to Add an Instance Annotation to a property
+        /// </summary>
+        /// <param name="propertyName">Name of the property</param>
+        /// <param name="annotationName">Name of Annotation</param>
+        /// <param name="value">Value of Annotation</param>
+        void AddPropertyAnnotation(string propertyName, string annotationName, object value);
+
+        /// <summary>
+        /// Get an Instance Annotation from CLR Type
+        /// </summary>
+        /// <param name="annotationName">Name of Annotation</param>
+        /// <returns>Get Annotation value for the given annotation</returns>
+        object GetResourceAnnotation(string annotationName);
+
+        /// <summary>
+        /// Get an Instance Annotation from the Property
+        /// </summary>
+        /// <param name="propertyName">Name of the Property</param>
+        /// <param name="annotationName">Name of the Annotation</param>
+        /// <returns>Get Annotation value for the given annotation and property</returns>
+        object GetPropertyAnnotation(string propertyName, string annotationName);
+
+        /// <summary>
+        /// Get All Annotations from CLR Type
+        /// </summary>
+        /// <returns>Dictionary of string(annotation name) and object value(annotation value)</returns>
+        IDictionary<string,object> GetAllResourceAnnotations();
+
+        /// <summary>
+        /// Get all Annotations for a Property
+        /// </summary>
+        /// <param name="propertyName">Name of Property</param>
+        /// <returns>Dictionary of string(annotation name) and object value(annotation value)</returns>
+        IDictionary<string, object> GetAllPropertyAnnotations(string propertyName);
```

Comment:
Maybe better to only include a property:  

IDictionary<string, IDictionary<string, object>> instanceAnnotations

in this interface?

Reply:
Mike Suggested its better we add and get instanceannotations using methods, @mikepizzo  Can you please comment

Reply:
Yes; it seemed simpler to me to have an interface that added/retrieved resource or property annotations, and leave the implementation details (i.e., use of empty string for properties on the type) as internals to the class.  I think this is cleaner, although I do realize now that there's not a way to *remove* an instance annotation.

An alternative would be to have IODataInstanceAnnotationContainer derive from IDictionary<string,IDictionary<string,object>>.  Then we could have the helper methods but the developer could still get to the underlying dictionaries (i.e., to remove an entry).

---
In reply to: [488995806](https://github.com/OData/WebApi/pull/2219#discussion_r488995806) [](ancestors = 488995806)